v1 addon with gts + ember-template-imports failing under Embroider

When a v1 addon has gts components and ember-template-imports, this is working under Classic, but failing under Embroider (stable). Imports of components used in `<template>` get stripped away (in the rewritten package), thereby leading to `Attempted to invoke a component that was not in scope in a strict mode template` errors.

Found some prior discussion around that on [Discord](https://discord.com/channels/480462759797063690/568935504288940056/1237704777211842600) between @Windvis and @NullVoxPopuli, where it was suggested to make the v1 babel typescript config same as for v2 addons. Which @Windvis did [here](https://github.com/appuniversum/ember-appuniversum/pull/492/files#r1599636132), and I can confirm in my own case that adding `onlyRemoveTypeImports` fixes the issue.

However, we should get this to work out of the box, right? It was suggested to add this config to ember-cli-babel by default, but not sure if that is viable? Couldn't this break stuff when imports are only used for types, but not declared explicitly with e.g. `import type`?

Also puzzling to me is the fact that this only causes errors under Embroider? (which is the reason I opened this issue here, even when the fix might go somewhere else)

hi @patricklx 

Is that added to the babel section of `ember-cli-build.js`? And if so, I'm getting an "Cannot find module './babel-plugin-ember-template-compilation.js'" error.

--- a/packages/compat/src/v1-addon.ts
+++ b/packages/compat/src/v1-addon.ts
@@ -116,9 +116,10 @@ export default class V1Addon {
   @Memoize()
   private get templateCompilerBabelPlugin(): PluginItem | undefined {
     let plugins = loadAstPlugins(this.addonInstance.registry);
+    let hasTemplateTag = this.addonInstance.addons.find((a: any) => a.name === 'ember-template-imports');
     // our macros don't run here in stage1
     plugins = plugins.filter((p: any) => !isEmbroiderMacrosPlugin(p));
-    if (plugins.length > 0) {
+    if (plugins.length > 0 || hasTemplateTag) {
       let compilerPath = require.resolve('ember-source/dist/ember-template-compiler.js', {
         paths: [findTopmostAddon(this.addonInstance).parent.root],
       });
@@ -203,6 +204,34 @@ export default class V1Addon {
     }
     if (this.addonInstance.addons.find((a: any) => a.name === 'ember-cli-htmlbars-inline-precompile')) {
       // the older inline template compiler is present
+      return true;
+    }
+    if (this.addonInstance.addons.find((a: any) => a.name === 'ember-template-imports')) {
+      /**
+       * Stage1 will always run custom broccoli preprocessors. So that's enough to convert:
+       *
+       * import Thing from './thing';
+       * <template><Thing/></template>
+       * to
+       *
+       * import Thing from './thing';
+       * import { template } from '@ember/template-compiler';
+       * export default template("Thing", {
+       *   eval: function() { return eval(arguments[0]) } })
+       * });
+       * This is really all we need to do at stage1, since this is now valid Javascript that could appear in a v2 addon.
+       *
+       * But if the addon is also using TS, we also need to run the typescript transform before it will be valid JS. And if the typescript transform was being truly correct it would not try to delete the import because the eval can see the imported binding. That's why we have an eval. It's a standards-compliant want of gaining access to everything in scope.
+       *
+       * Normally we only use babel-plugin-ember-template-compilation in stage1 to run custom AST transforms. Since there are none in the addon, we don't add it. The fix here is helping because there is a new reason to add it. It will further convert the above example to:
+       *
+       * import Thing from './thing';
+       * import { template } from '@ember/template-compiler';
+       * export default template("Thing", {
+       *   scope: () => ({ Thing })
+       * });
+       * which typescript then respects.
+       */
       return true;
     }
 

--- a/tests/scenarios/compat-stage2-test.ts
+++ b/tests/scenarios/compat-stage2-test.ts
@@ -254,6 +254,86 @@ stage2Scenarios
   });
 
 stage2Scenarios
+  .map('gts-files-in-addons-are-pre-processed-with-template-compilation', app => {
+    let depA = addAddon(app, 'dep-a');
+    depA.linkDependency('ember-template-imports', { baseDir: __dirname });
+    depA.linkDependency('ember-cli-babel', { baseDir: __dirname, resolveName: 'ember-cli-babel-latest' });
+
+    merge(depA.files, {
+      'index.js': `
+        'use strict';
+        module.exports = {
+          name: require('./package').name,
+          options: {
+            'ember-cli-babel': { enableTypeScriptTransform: true },
+          },
+        };`,
+      addon: {
+        components: {
+          'other.gts': `
+          import Component from '@glimmer/component';
+
+          export default class extends Component {
+            abc: string;
+            <template>
+              other
+            </template>
+          };
+          `,
+          'gts-component.gts': `
+          import Component from '@glimmer/component';
+          import OtherComponent from './other';
+
+          export default class extends Component {
+            abc: string;
+            <template>
+              this is gts
+              with <OtherComponent />
+            </template>
+          };
+          `,
+        },
+      },
+      app: {
+        components: {
+          'gts-component.js': 'export { default } from "dep-a/components/gts-component"',
+        },
+      },
+    });
+  })
+  .forEachScenario(scenario => {
+    Qmodule(scenario.name, function (hooks) {
+      throwOnWarnings(hooks);
+
+      let app: PreparedApp;
+
+      hooks.before(async assert => {
+        app = await scenario.prepare();
+        let result = await app.execute('ember build', { env: { STAGE2_ONLY: 'true' } });
+        assert.equal(result.exitCode, 0, result.output);
+      });
+
+      let expectAudit = setupAuditTest(hooks, () => ({ app: app.dir, 'reuse-build': true }));
+
+      test('no audit issues', function () {
+        expectAudit.hasNoFindings();
+      });
+
+      test('gts is processed with template-compilation', function () {
+        let expectModule = expectAudit.module('./assets/my-app.js');
+        // this is to make sure that the babel plugin template compilation runs and thus
+        // make imports that are only used in templates bound and not removed by typescript
+        expectModule
+          .resolves('my-app/components/gts-component.js')
+          .toModule()
+          .resolves('dep-a/components/gts-component')
+          .toModule()
+          .codeContains(`import OtherComponent from './other';`);
+      });
+    });
+  });
+
+stage2Scenarios
   .map('static-with-rules', app => {
     app.addDependency('some-library', '1.0.0');
     app.linkDependency('@embroider/sample-transforms', { baseDir: __dirname });

--- a/tests/scenarios/package.json
+++ b/tests/scenarios/package.json
@@ -94,6 +94,7 @@
     "ember-source-canary": "https://s3.amazonaws.com/builds.emberjs.com/canary/shas/756f0e3f98b8ca5edf443fe57318b4dac692bffa.tgz",
     "ember-source-latest": "npm:ember-source@latest",
     "ember-truth-helpers": "^3.0.0",
+    "ember-template-imports": "^4.1.2",
     "execa": "^5.1.1",
     "popper.js": "^1.16.1",
     "tslib": "^2.6.0",

